cli/config_file.go

```diff
@@ -384,3 +390,88 @@ func RandomChars(bufflength uint) string {
 
 	return result[0:bufflength]
 }
+
+const (
+	CruxAgentGrpcPort  = "CruxAgentGrpcPort"
+	CruxGrpcPort       = "CruxGrpcPort"
+	CruxUIPort         = "CruxUIPort"
+	KratosAdminPort    = "KratosAdminPort"
+	KratosPublicPort   = "KratosPublicPort"
+	KratosPostgresPort = "KratosPostgresPort"
+	MailSlurperPort    = "MailSlurperPort"
+	MailSlurperPort2   = "MailSlurperPort2"
+	CruxPostgresPort   = "CruxPostgresPort"
+)
+
+func CheckAndUpdatePorts(settings *Settings) *Settings {
+	portMap := map[string]uint{}
+	portMap[CruxAgentGrpcPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.CruxAgentGrpcPort, CruxAgentGrpcPort)
+	portMap[CruxGrpcPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.CruxGrpcPort, CruxGrpcPort)
+	portMap[CruxUIPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.CruxUIPort, CruxUIPort)
+	portMap[CruxPostgresPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.CruxPostgresPort, CruxPostgresPort)
+	portMap[KratosAdminPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.KratosAdminPort, KratosAdminPort)
+	portMap[KratosPublicPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.KratosPublicPort, KratosPublicPort)
+	portMap[KratosPostgresPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.KratosPostgresPort, KratosPostgresPort)
+	portMap[MailSlurperPort] = GetAvailablePort(portMap, settings.SettingsFile.Options.MailSlurperPort, MailSlurperPort)
+	portMap[MailSlurperPort2] = GetAvailablePort(portMap, settings.SettingsFile.Options.MailSlurperPort2, MailSlurperPort2)
+
+	settings.SettingsFile.Options.CruxAgentGrpcPort = portMap[CruxAgentGrpcPort]
+	settings.SettingsFile.Options.CruxGrpcPort = portMap[CruxGrpcPort]
+	settings.SettingsFile.Options.CruxUIPort = portMap[CruxUIPort]
+	settings.SettingsFile.Options.CruxPostgresPort = portMap[CruxPostgresPort]
+	settings.SettingsFile.Options.KratosAdminPort = portMap[KratosAdminPort]
+	settings.SettingsFile.Options.KratosPublicPort = portMap[KratosPublicPort]
+	settings.SettingsFile.Options.KratosPostgresPort = portMap[KratosPostgresPort]
+	settings.SettingsFile.Options.MailSlurperPort = portMap[MailSlurperPort]
+	settings.SettingsFile.Options.MailSlurperPort2 = portMap[MailSlurperPort2]
+
+	return settings
+}
+
+func GetAvailablePort(portMap map[string]uint, portNum uint, portDesc string) uint {
+	p := portNum
```

Comment:
Any reason why you add a new variable `p`?

Reply:
no reason, removed `p` variable